### PR TITLE
[BugFix] Fix bug the user still has permissions after revoke role from user

### LINKED ISSUE #17885: [access control]After revoke role from user, the user still has permissions
### Steps to reproduce the behavior (Required)
step 1
conn1: mysql -hxxx.xx.xx.xxx -P9232 -uroot
```
conn1: mysql -hxxx.xx.xx.xxx -P9232 -uroot
create user user_aaaaaaa identified by '123456';
create role role_bbbbbbb ;
GRANT create_external_catalog ON SYSTEM TO role role_bbbbbbb ;
GRANT role_bbbbbbb TO user_aaaaaaa ;
 alter user user_aaaaaaa default role role_bbbbbbb ;
```
step 2
conn2: mysql -hxxx.xx.xx.xxx -P9232 -uuser_aaaaaaa -p123456
```
CREATE EXTERNAL CATALOG catalog_111111 properties ("type" = "hive","hive.metastore.uris" = "thrift://xxx.xx.xx.xxx:9083") ;
```
step 3
conn1:
```
REVOKE role_bbbbbbb FROM user_aaaaaaa ;
show grants for user_aaaaaaa;
```
step 4
conn2:
```
CREATE EXTERNAL CATALOG catalog_222222 properties ("type" = "hive","hive.metastore.uris" = "thrift://xxx.xx.xx.xxx:9083") ;
```
step 5
conn3:mysql -hxxx.xx.xx.xxx -P9232 -uuser_aaaaaaa -p123456
```
CREATE EXTERNAL CATALOG catalog_333333 properties ("type" = "hive","hive.metastore.uris" = "thrift://xxx.xx.xx.xxx:9083") ;

```

image

image

image

### Expected behavior (Required)
step4 and step 5 failed
### Real behavior (Required)
success
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
mysql> select current_version();
+----------------------+
| current_version()    |
+----------------------+
| MAIN-RELEASE ec3160e |
+----------------------+
1 row in set (0.00 sec)

### PR DESCRIPTION
**Pull request by HangyuanLiu**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #17885

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 1]
```diff
@@ -1233,11 +1233,11 @@ executeAsStatement
     ;
 
 createRoleStatement
-    : CREATE ROLE (IF NOT EXISTS)? identifierOrString
+    : CREATE ROLE (IF NOT EXISTS)? roleList
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:1236]: why we need to create multi roles on time?

Comment (HangyuanLiu) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:1236]: This is not a bugfix, just a new mini feature. We don't have to support it, but we can provide this ability.

File: fe/fe-core/src/main/java/com/starrocks/qe/SetDefaultRoleExecutor.java [hunk 2]
```diff
@@ -40,10 +39,9 @@ public static void execute(SetDefaultRoleStmt stmt, ConnectContext context) thro
             roleIds = new HashSet<>();
             for (String roleName : stmt.getRoles()) {
                 Long roleId = manager.getRoleIdByNameAllowNull(roleName);
-                if (roleId == null || !roleIdsForUser.contains(roleId)) {
-                    throw new SemanticException("Role " + roleName + " is not granted to " + user.toString());
+                if (roleId != null && roleIdsForUser.contains(roleId)) {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/qe/SetDefaultRoleExecutor.java:42]: So, we won't prompt an error if user set an unowned role as his default role, why we change this logic?

Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/qe/SetDefaultRoleExecutor.java:42]: got it, it's moved to the analyisis phase

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PrivilegeCheckerV2.java [hunk 3]
```diff
@@ -1098,21 +1100,38 @@ public Void visitShowRolesStatement(ShowRolesStmt statement, ConnectContext cont
         }
 
         @Override
-        public Void visitGrantRevokePrivilegeStatement(BaseGrantRevokePrivilegeStmt stmt, ConnectContext context) {
-            PrivilegeManager privilegeManager = context.getGlobalStateMgr().getPrivilegeManager();
-            if (!privilegeManager.allowGrant(context, stmt.getObjectType(), stmt.getPrivilegeTypes(), stmt.getObjectList())) {
+        public Void visitGrantRoleStatement(GrantRoleStmt statement, ConnectContext context) {
+            if (statement.getGranteeRole().stream().anyMatch(
+                    r -> r.equalsIgnoreCase("root") || r.equalsIgnoreCase("cluster_admin"))) {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PrivilegeCheckerV2.java:1105]: for "root" and "cluster_admint", better to use a constant in PrivilegeManager

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PrivilegeCheckerV2.java [hunk 4]
```diff
@@ -1098,21 +1100,38 @@ public Void visitShowRolesStatement(ShowRolesStmt statement, ConnectContext cont
...
             if (statement.getGranteeRole().stream().anyMatch(
                     r -> r.equalsIgnoreCase("root") || r.equalsIgnoreCase("cluster_admin"))) {
                 UserIdentity userIdentity = context.getCurrentUserIdentity();
                 if (!userIdentity.equals(UserIdentity.ROOT)) {
-                    throw new SemanticException("Can not grant/revoke root or cluster_admin role except root user");
+                    throw new SemanticException("Can not grant root or cluster_admin role except root user");
+                }
+            }
+
+            if (statement.getGranteeRole().stream().anyMatch(r -> r.equalsIgnoreCase("root"))) {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PrivilegeCheckerV2.java:1132]: we should only limit that : we can't revoke root role from root?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 180 / 235 (76.60%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|10|20|50.00%|[1887, 2018, 2040, 2041, 2042, 2043, 2045, 2057, 2058, 2059]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeCheckerV2.java|11|21|52.38%|[1121, 1128, 1133, 1134, 1146, 1147, 1148, 1149, 1151, 1158]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|3|66.67%|[1431]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeManager.java|110|142|77.46%|[1401, 1403, 1404, 1405, 1406, 1407, 1408, 1410, 1437, 1459, 1460, 1471, 1477, 1478, 1496, 1497, 1498, 1499, 1501, 1504, 1505, 1511, 1512, 1524, 1525, 1532, 1534, 1535, 1536, 1537, 1538, 1540]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|17|19|89.47%|[433, 434]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantRoleStmt.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokeRoleStmt.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/UserPrivilegeCollection.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/RolePrivilegeCollectionInfo.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateRoleStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SetDefaultRoleExecutor.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/DropRoleStmt.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 221 / 247 (89.47%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|3|66.67%|[1431]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeCheckerV2.java|13|17|76.47%|[1106, 1128, 1147, 1157]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeManager.java|129|146|88.36%|[1439, 1461, 1462, 1473, 1479, 1480, 1498, 1499, 1500, 1501, 1503, 1506, 1507, 1513, 1514, 1526, 1527]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|19|21|90.48%|[433, 434]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|18|20|90.00%|[1887, 2018]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantRoleStmt.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokeRoleStmt.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/UserPrivilegeCollection.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AstToStringBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/RolePrivilegeCollectionInfo.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateRoleStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SetDefaultRoleExecutor.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokeRoleStmt.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/DropRoleStmt.java|3|3|100.00%|[]|

**Review by Astralidea — DISMISSED** [review]

**Review by gengjun-git — DISMISSED** [review]

**Comment by HangyuanLiu** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java [hunk 5]
```diff
@@ -1274,20 +1293,41 @@ protected PrivilegeCollection loadPrivilegeCollection(UserIdentity userIdentity,
...
             try {
                 // 1. get all parent roles by default, but can be specified with `SET ROLE` statement
                 if (roleIds == null) {
-                    roleIds = userCollection.getAllRoles();
+                    roleIds = userPrivilegeCollection.getAllRoles();
                 }
 
                 // 2. get all predecessors base on step 1
-                roleIds = getAllPredecessorsUnlocked(roleIds);
+                Set<Long> parentRoleIds = new HashSet<>();
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java:1306]: before find all the parent role ids, we should remove the role id from `rolesIds` set which doesn't belong to this user any more?

File: fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java [hunk 6]
```diff
@@ -1380,11 +1431,119 @@ public RolePrivilegeCollection getRolePrivilegeCollectionUnlocked(long roleId, b
...
+                    getRolePrivilegeCollectionUnlocked(roleId, true);
+            return rolePrivilegeCollection.getTypeToPrivilegeEntryList();
+        } catch (PrivilegeException e) {
+            throw new SemanticException(e.getMessage());
+        } finally {
+            roleReadUnlock();
+        }
+    }
+
+    public List<String> getGranteeRoleToUser(UserIdentity userIdentity) {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java:1486]: maybe better to rename to `getGranteeRoleDetailsForUser()`

**Comment by HangyuanLiu** [comment]
run starrocks_admit_test

**Review by nshangyiming — APPROVED** [review]

**Comment by HangyuanLiu** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 214 / 239 (89.54%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeCheckerV2.java|13|17|76.47%|[1106, 1128, 1147, 1157]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeManager.java|124|141|87.94%|[1428, 1450, 1451, 1462, 1468, 1469, 1487, 1488, 1489, 1490, 1492, 1495, 1496, 1502, 1503, 1515, 1516]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|19|21|90.48%|[433, 434]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|18|20|90.00%|[1887, 2018]|
|:large_blue_circle:|com/starrocks/sql/ast/GrantRoleStmt.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RevokeRoleStmt.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/UserPrivilegeCollection.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AstToStringBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/mysql/privilege/Auth.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/persist/RolePrivilegeCollectionInfo.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateRoleStmt.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SetDefaultRoleExecutor.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzer.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/BaseGrantRevokeRoleStmt.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/DropRoleStmt.java|3|3|100.00%|[]|

**Comment by HangyuanLiu** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**